<center>
    <h1><b>Prediction of Bank Marketing Responses</b></h1>
</center>

**Table of Contents:**
    

  1.&nbsp;[Problem Statement and Background Information](#section1)<br>
         
  2.&nbsp;[Data Loading](#section2)<br>
       
  3.&nbsp;[Data Analysis and  pre-processing](#section3)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1&nbsp;[Output Variable Analysis](#section31)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2&nbsp;[IQR and data filtering using IQR](#section32)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.3&nbsp;[Handling Categorical variable through Label Encoding](#section33)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.4&nbsp;[Train-Test Split](#section34)<br>
      
  4.&nbsp;[Data modeling](#section4)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.1&nbsp;[Building Logistic Regression Model](#section41)<br>
      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4.2&nbsp;[Model Evaluation](#section42)<br>
      
      
  5.&nbsp;[Conclusions](#section5)<br>
  


<a id='section1'></a>
<h2><b> 1. Problem Statement and Background information</b></h2>

**Background Information:**
The data is related with direct marketing campaigns of a Portuguese banking institution. The marketing campaigns were based on phone calls. Often, more than one contact to the same client was required, in order to access if the __product (bank term deposit)__ would be __('yes')__ or __('no')__ subscribed. 

**Problem Statement** 
The data is related with __direct marketing campaigns__ (phone calls) of a Portuguese banking institution. The classification goal is to predict if the client will subscribe a __term deposit (variable y)__.



<h3>Source:</h3>

 - Dataset from : http://archive.ics.uci.edu/ml/datasets/Bank+Marketing#
 

<a id="section2"></a>
<h2><b>2. Data Loading</b></h2>


In [0]:
import pandas as pd
bank = pd.read_csv('https://raw.githubusercontent.com/insaid2018/Term-2/master/Data/bank.csv')
bank.head() 

age          job  marital          education  default housing loan  \
0   46  blue-collar  married           basic.9y       no      no  yes   
1   29       admin.   single  university.degree       no      no   no   
2   50   management  married  university.degree       no     yes  yes   
3   31       admin.  married        high.school  unknown     yes   no   
4   32       admin.   single        high.school       no     yes   no   

     contact month day_of_week ...   campaign  pdays  previous     poutcome  \
0  telephone   may         tue ...          3    999         0  nonexistent   
1   cellular   may         wed ...          2    999         0  nonexistent   
2   cellular   apr         fri ...          1    999         0  nonexistent   
3  telephone   may         thu ...          2    999         0  nonexistent   
4   cellular   jun         tue ...          1    999         0  nonexistent   

  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed    y  
0          1.1          93.994          -36.4      4.857       5191.0   no  
1         -1.8          92.893          -46.2      1.281       5099.1   no  
2         -1.8          93.075          -47.1      1.405       5099.1   no  
3          1.1          93.994          -36.4      4.860       5191.0   no  
4         -2.9          92.963          -40.8      1.262       5076.2  yes  

[5 rows x 21 columns]

<a id="section3"></a>
<h2><b>3. Data Analysis and Preprocessing</b></h2>

### Check the columns present in the dataset

In [0]:
bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

### Check the shape of Dataset

In [0]:
bank.shape 

(10297, 21)

### Check the descriptive statistics of the dataset

In [0]:
bank.describe() 

age      duration      campaign         pdays      previous  \
count  10297.000000  10297.000000  10297.000000  10297.000000  10297.000000   
mean      40.080606    261.388268      2.561134    963.609692      0.171506   
std       10.472190    263.722874      2.827084    184.098592      0.496992   
min       18.000000      0.000000      1.000000      0.000000      0.000000   
25%       32.000000    103.000000      1.000000    999.000000      0.000000   
50%       38.000000    180.000000      2.000000    999.000000      0.000000   
75%       47.000000    327.000000      3.000000    999.000000      0.000000   
max       94.000000   3643.000000     56.000000    999.000000      6.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  10297.000000    10297.000000   10297.000000  10297.000000  10297.000000  
mean       0.077256       93.574206     -40.445780      3.621923   5166.850442  
std        1.576732        0.577502       4.622221      1.734170     72.492914  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

### Check the info of the dataset

In [0]:
bank.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10297 entries, 0 to 10296
Data columns (total 21 columns):
age               10297 non-null int64
job               10297 non-null object
marital           10297 non-null object
education         10297 non-null object
default           10297 non-null object
housing           10297 non-null object
loan              10297 non-null object
contact           10297 non-null object
month             10297 non-null object
day_of_week       10297 non-null object
duration          10297 non-null int64
campaign          10297 non-null int64
pdays             10297 non-null int64
previous          10297 non-null int64
poutcome          10297 non-null object
emp.var.rate      10297 non-null float64
cons.price.idx    10297 non-null float64
cons.conf.idx     10297 non-null float64
euribor3m         10297 non-null float64
nr.employed       10297 non-null float64
y                 10297 non-null object
dtypes: float64(5), int64(5), object(11)
memory usa

### Code to understand the below mentioned variables and print their 'unique' attributes.
``` ['job', 'marital', 'education', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'campaign','poutcome', 'pdays', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m'] ```

In [0]:
def unique_values():
    columns=['job', 'marital', 'education', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'campaign','poutcome', 'pdays', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m']
    for col in columns:
        print("Column Name: {0} , Unique values: {1}".format(col,len(bank[col].unique())))
        print(bank[col].unique())
unique_values()

Column Name: job , Unique values: 12
['blue-collar' 'admin.' 'management' 'technician' 'retired' 'services'
 'entrepreneur' 'self-employed' 'unemployed' 'student' 'housemaid'
 'unknown']
Column Name: marital , Unique values: 4
['married' 'single' 'divorced' 'unknown']
Column Name: education , Unique values: 8
['basic.9y' 'university.degree' 'high.school' 'professional.course'
 'basic.6y' 'basic.4y' 'unknown' 'illiterate']
Column Name: housing , Unique values: 3
['no' 'yes' 'unknown']
Column Name: loan , Unique values: 3
['yes' 'no' 'unknown']
Column Name: contact , Unique values: 2
['telephone' 'cellular']
Column Name: month , Unique values: 10
['may' 'apr' 'jun' 'aug' 'jul' 'oct' 'nov' 'dec' 'sep' 'mar']
Column Name: day_of_week , Unique values: 5
['tue' 'wed' 'fri' 'thu' 'mon']
Column Name: campaign , Unique values: 36
[ 3  2  1  4  5  6 14  7 21  9 12  8 17 10 20 33 29 15 23 11 18 13 19 39
 27 16 31 34 26 28 24 32 22 56 25 40]
Column Name: poutcome , Unique values: 3
['nonexistent' 

### Code to check the min and max value of age. Also check if there is any null value or not.

In [0]:
def age_analysis():
    # print the max age
    # print the min age
    # print the missing values if any
    print("Youngest member age:",bank['age'].min())
    print("Oldest member age:",bank['age'].max())
    print("Error values(Null):",bank['age'].isnull().sum())
age_analysis()

Youngest member age: 18
Oldest member age: 94
Error values(Null): 0


<a id="section31"> </a>

<h3><b>3.1 Output variable Analysis</b></h3>

### Count of Yes and No for the term deposit 

In [0]:
bank.y.value_counts()

no     9137
yes    1160
Name: y, dtype: int64

<a id="section32"></a>
<h3><b>3.2 IQR and data filtering based on IQR</b><h3>

### Function to calculate the Inter quartile range for quantile values outside 25 to 75 range. And to do the outlier capping for lower level with min value and for upper level with 'q3=1.5*iqr' value.

In [0]:
def remove_outlier(df_in, col_name):
    Q1=df_in[col_name].quantile(0.25)
    Q3=df_in[col_name].quantile(0.75)
    # Calculating IQR as difference of Quantile 3 and quantile 1
    IQR=Q3-Q1
    min_col=df_in.min()
    max_col=df_in.max()
    lower_bound=df_in[col_name].min()
    upper_bound=Q3 + 1.5*IQR
    print('Column',col_name,'IQR lower bound and upper bound are', lower_bound, 'and', upper_bound, 'respectively')
    df_out=df_in[(df_in[col_name]>lower_bound) & (df_in[col_name]<upper_bound)]
    return df_out

### Using the above created function , remove the outlier from 'age' variables:

In [0]:
def age_outlier(bank):
    # your code to remove the outlier from age
    col_name='age'
    df=remove_outlier(bank,col_name)
    return df
    
bank = age_outlier(bank) 

Column age IQR lower bound and upper bound are 18 and 69.5 respectively


In [0]:
print(bank)

       age            job   marital            education  default  housing  \
0       46    blue-collar   married             basic.9y       no       no   
1       29         admin.    single    university.degree       no       no   
2       50     management   married    university.degree       no      yes   
3       31         admin.   married          high.school  unknown      yes   
4       32         admin.    single          high.school       no      yes   
5       33         admin.    single    university.degree       no      yes   
6       51     management   married    university.degree       no      yes   
7       40    blue-collar   married             basic.9y       no       no   
8       34    blue-collar   married  professional.course  unknown       no   
9       35         admin.    single          high.school       no       no   
10      55         admin.   married          high.school       no  unknown   
11      34     technician   married  professional.course       n

### Using the above created function , remove the outlier from 'campaign' variables:

In [0]:
def campaign_outlier(bank):
    col_name='campaign'
    df=remove_outlier(bank,col_name)
    return df

    
bank = campaign_outlier(bank) 

Column campaign IQR lower bound and upper bound are 1 and 6.0 respectively


### Using the above created function , remove the outlier from 'duration' variables:

In [0]:
def duration_outlier(bank):
    col_name='duration'
    df=remove_outlier(bank,col_name)
    return df
    
bank = duration_outlier(bank) 

Column duration IQR lower bound and upper bound are 0 and 684.0 respectively


### Dividing dataset into two, on the basis of categorical and numerical.

In [0]:
bank_cat=bank[['job', 'marital','default', 'education', 'loan', 'housing', 'contact', 'month', 'day_of_week', 'poutcome', 'y']]

In [0]:
bank_cont = bank.drop(['job', 'marital','default', 'education', 'loan', 'housing', 'contact', 'month', 'day_of_week', 'poutcome', 'y'], axis=1)

<a id="section33"></a>
<h3><b>3.3 Handling categorial variables through Label Encoding</b></h3>

###  Label encoded the below mentioned categorical variable to numerical values.
``` ['job', 'marital','default','education', 'loan', 'housing', 'contact', 'month', 'day_of_week', 'poutcome', 'y'] ```

In [0]:
from sklearn.preprocessing import LabelEncoder
def label_encode(bank_cat):
    # code to return the value applying fit_transorm
    return bank_cat.apply(LabelEncoder().fit_transform)
bank_cat = label_encode(bank_cat)

### Combining the numerical and categorical dataset.

In [0]:
bank_final= pd.concat([bank_cont, bank_cat], axis = 1)

<a href="section34"></a>
<h3><b>3.4 Train-Test Split</b></h3>

### Extract independent column to prepare X 

In [0]:
X = pd.DataFrame()
def feature_selection2(): #All columns except the predictor column in sales only few were selected
    # write your code to create a dataframe of dependent variables excluding 'y' variable
    X = bank_final.loc[:,bank.columns != 'y']
    return X
X = feature_selection2()

In [0]:
bank1=bank.columns.sort_values()
bank2=bank_final.columns.sort_values()

In [0]:
bank1==bank2

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True])

In [0]:
bank_final

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       46       329         3    999         0           1.1          93.994   
1       29       132         2    999         0          -1.8          92.893   
3       31       199         2    999         0           1.1          93.994   
5       33       174         4    999         0           1.4          93.444   
8       34        65         2    999         1          -1.8          92.893   
9       35       282         5    999         0          -1.8          92.893   
11      34       211         2      6         2          -1.8          93.749   
14      54       225         2    999         0           1.4          93.918   
18      53       340         2    999         0           1.4          94.465   
25      52       112         2    999         0           1.4          94.465   
26      51       176         3    999         0           1.1          93.994   
29      35       207         2    999         0           1.4          93.918   
34      33       409         2    999         0          -1.8          92.893   
35      52       434         2    999         0          -1.8          92.893   
37      32       396         3    999         0           1.4          93.918   
44      27        64         2    999         0           1.4          94.465   
46      45       126         2    999         1          -0.1          93.200   
48      59       133         3    999         0           1.4          93.918   
51      35       113         2    999         0           1.4          94.465   
55      59       251         5    999         0           1.4          93.918   
56      39        90         3    999         0          -0.1          93.200   
57      58       275         4    999         0           1.4          93.918   
59      39        99         5    999         0           1.4          93.918   
61      37       574         5    999         0           1.1          93.994   
62      47       374         2     16         1          -1.7          94.027   
63      29        92         3    999         0           1.1          93.994   
66      52       291         3    999         0           1.4          94.465   
68      35       541         2    999         0           1.1          93.994   
71      46        76         3    999         0          -0.1          93.200   
77      32        86         3    999         0          -0.1          93.200   
...    ...       ...       ...    ...       ...           ...             ...   
10234   31       406         2    999         0           1.4          94.465   
10236   46       279         4    999         0           1.4          93.918   
10241   59       133         2    999         0           1.4          94.465   
10242   52       402         2    999         0           1.4          93.918   
10247   58       483         2    999         0          -1.8          93.075   
10249   38       121         4    999         0          -1.8          93.075   
10250   39       433         3    999         0           1.1          93.994   
10253   41       134         2    999         0           1.4          93.918   
10254   34       295         3    999         0           1.4          93.918   
10256   45       220         4    999         1          -0.1          93.200   
10260   28       128         2    999         0           1.1          93.994   
10262   56       162         3    999         0          -0.1          93.200   
10264   48       349         2    999         0           1.4          94.465   
10265   38       136         2    999         1          -0.1          93.200   
10266   56       150         2    999         0           1.4          94.465   
10268   54       150         2    999         0           1.4          93.918   
10270   52       121         3    999         0           1.4          93.444   
10271   41       271         3    999 

###  Extract dependent variable into a dataframe 'y' for model predcition

In [0]:
y = pd.DataFrame()
def target_variable():
    # write your code to create a dataframe which consists only of dependepent variable
    y=bank_final['y']
    return y
y = target_variable()

### Splitting X and y intro train and test dataset

In [0]:
X

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       46       329         3    999         0           1.1          93.994   
1       29       132         2    999         0          -1.8          92.893   
3       31       199         2    999         0           1.1          93.994   
5       33       174         4    999         0           1.4          93.444   
8       34        65         2    999         1          -1.8          92.893   
9       35       282         5    999         0          -1.8          92.893   
11      34       211         2      6         2          -1.8          93.749   
14      54       225         2    999         0           1.4          93.918   
18      53       340         2    999         0           1.4          94.465   
25      52       112         2    999         0           1.4          94.465   
26      51       176         3    999         0           1.1          93.994   
29      35       207         2    999         0           1.4          93.918   
34      33       409         2    999         0          -1.8          92.893   
35      52       434         2    999         0          -1.8          92.893   
37      32       396         3    999         0           1.4          93.918   
44      27        64         2    999         0           1.4          94.465   
46      45       126         2    999         1          -0.1          93.200   
48      59       133         3    999         0           1.4          93.918   
51      35       113         2    999         0           1.4          94.465   
55      59       251         5    999         0           1.4          93.918   
56      39        90         3    999         0          -0.1          93.200   
57      58       275         4    999         0           1.4          93.918   
59      39        99         5    999         0           1.4          93.918   
61      37       574         5    999         0           1.1          93.994   
62      47       374         2     16         1          -1.7          94.027   
63      29        92         3    999         0           1.1          93.994   
66      52       291         3    999         0           1.4          94.465   
68      35       541         2    999         0           1.1          93.994   
71      46        76         3    999         0          -0.1          93.200   
77      32        86         3    999         0          -0.1          93.200   
...    ...       ...       ...    ...       ...           ...             ...   
10234   31       406         2    999         0           1.4          94.465   
10236   46       279         4    999         0           1.4          93.918   
10241   59       133         2    999         0           1.4          94.465   
10242   52       402         2    999         0           1.4          93.918   
10247   58       483         2    999         0          -1.8          93.075   
10249   38       121         4    999         0          -1.8          93.075   
10250   39       433         3    999         0           1.1          93.994   
10253   41       134         2    999         0           1.4          93.918   
10254   34       295         3    999         0           1.4          93.918   
10256   45       220         4    999         1          -0.1          93.200   
10260   28       128         2    999         0           1.1          93.994   
10262   56       162         3    999         0          -0.1          93.200   
10264   48       349         2    999         0           1.4          94.465   
10265   38       136         2    999         1          -0.1          93.200   
10266   56       150         2    999         0           1.4          94.465   
10268   54       150         2    999         0           1.4          93.918   
10270   52       121         3    999         0           1.4          93.444   
10271   41       271         3    999 

In [0]:
from sklearn.model_selection import train_test_split
def df_split():
    
    return train_test_split(X, y , test_size=0.25, random_state=1)# train test split using train_test_split of 75:25 and random state=1
X_train, X_test, y_train, y_test = df_split()

### Check the shape of X an y of train dataset.

In [0]:
def shape_check_train():
    print(X_train.shape)
    print(y_train.shape)
shape_check_train()

(3465, 20)
(3465,)


### Check the shape of X and y of test dataset.

In [0]:
def shape_check_test():
    print(X_test.shape)
    print(y_test.shape)
shape_check_test()

(1156, 20)
(1156,)


<a id="section4"></a>
<h3><b>4. Data Modeling</b></h3>
<a id="section41"></a>
<h2><b>4.1 Building Logistic Regression Model</b></h2>

### Instantitating Logistic Regression model using scikit-learn

In [0]:
from sklearn.linear_model import LogisticRegression
def model_initialization():
    # initiate the logistic regression model to new variable logreg
    logreg=LogisticRegression()
    return logreg
logreg = model_initialization()

### Fitting the logistic model on X_train and y_train

In [0]:
def model_fit():
    # fit the X_train and y_train
    logreg.fit(X_train,y_train)
model_fit()    

### Using the model for prediction

In [0]:
y_pred_train = pd.DataFrame()
def logistic_prediction_train():
    # create the model prediction on X_train data using the above created dataframe
    y_pred_train=logreg.predict(X_train)
    return y_pred_train
y_pred_train = logistic_prediction_train()

In [0]:
y_pred_test = pd.DataFrame()
def logistic_prediction_test():
    # create the model prediction on X_test data using the above created dataframe
    y_pred_test=logreg.predict(X_test)
    return y_pred_test
y_pred_test = logistic_prediction_test()

<a id="section42"></a>

<h3><b>4.2 Model evaluation</b></h3>

### Model evaluation using accuracy classification score

In [0]:
from sklearn.metrics import accuracy_score
def logistic_accuracy_score():
    #Calculate and print the accuracy score
    print("The accuracy of test_prediction w.r.t actual_test is :",accuracy_score(y_test,y_pred_test))
    
logistic_accuracy_score()

The accuracy of test_prediction w.r.t actual_test is : 0.9403114186851211


### Model evaluation using Confusion matrix

In [0]:
from sklearn.metrics import confusion_matrix
confusion_matrix = pd.DataFrame(confusion_matrix(y_test, y_pred_test))

def model_evaluation():
    # provide confusion matrix index name as['Actual No_Deposit','Actual Deposit']
    # provide confusion matrix index name as['Predicted No_Deposit','Predicted Deposit']
    confusion_matrix.index=['Actual No_Deposit','Actual Deposit']
    confusion_matrix.columns=['Predicted No_Deposit', 'Predicted Deposit']
    print(confusion_matrix)
model_evaluation()

                   Predicted No_Deposit  Predicted Deposit
Actual No_Deposit                  1062                 14
Actual Deposit                       55                 25


### Accuracy prediction setting the threshold = 0.75

In [0]:
import numpy as np
def setting_threshold1():
    preds1 = np.where(logreg.predict_proba(X_test)[:,1]> 0.75,1,0)
    print('Accuracy score for test data is:', accuracy_score(y_test,preds1))
setting_threshold1()

Accuracy score for test data is: 0.9403114186851211


###  Accuracy prediction setting the threshold = 0.25
https://datascience.stackexchange.com/questions/22762/understanding-predict-proba-from-multioutputclassifier

In [0]:
def setting_threshold2():
    preds2 = np.where(logreg.predict_proba(X_test)[:,1]> 0.25,1,0)
    print('Accuracy score for test data is:', accuracy_score(y_test,preds2))

setting_threshold2()

Accuracy score for test data is: 0.9359861591695502
[0 0 0 ... 0 0 0]


<a id="section5"></a>
<h3><b>5. Conslusions</b></h3>

Using logistic regression model, the Bank Marketing call responses are classified with 94% accuracy.